### PR TITLE
DYT-3054: Add submit_batch API to MemoryLake

### PR DESCRIPTION
**Pull request by msrdic**
## Summary
- Adds `submit_batch()` to `MemoryLake`: accepts a list of document dicts, persists them as PENDING records atomically, then processes them in the background via a `asyncio.Task`
- Introduces `BatchHandle` (tracks progress, exposes `completed`/`failed`/`is_done`, and `await handle.wait()`) and `DocumentResult` (per-document outcome delivered to an optional `on_result` callback)
- Adds per-call `max_chunks_in_flight` override to `VectorCypherEngine.process_staged_document()` and `_process_document()`, taking precedence over the engine-level config
- Exports `BatchHandle` and `DocumentResult` from the top-level `khora` package (`__all__`)
- Adds `MemoryLake._bg_tasks` set to keep references to background tasks so they are not garbage-collected before completion

## Test plan
- [x] Unit tests pass (`tests/unit/test_memory_lake.py` — 435 lines of new tests covering happy path, partial failures, `on_result` callback, `wait()`, and edge cases)
- [x] Pre-existing failures in `test_logging_config.py` and `test_unique_external_id.py` are unrelated to this change (confirmed on `origin/main` baseline)
- [x] `make format` — no changes
- [x] Coverage ≥ 30% (52.81% total)

### CODE REVIEW

**Comment by msrdic** [comment]
## Audit Summary

| Audit | Status |
|-------|--------|
| Self-audit (`/audit --auto`) | clean (fixed in 2 cycles) |
| Specialist: `/review:test --auto` | clean (fixed in 2 cycles) |
| Bandit (Python security) | clean (fixed in 2 cycles) |
| Cross-audit (`/codex-audit`) | ran |

**Cycles:** 2/5
**Result:** clean (fixed in 2 cycles)

## Fix Cycle 1

### Files changed
```
src/khora/memory_lake.py       | 82 +++++++++++++++++++++++++++--------
 tests/unit/test_memory_lake.py | 98 ++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 162 insertions(+), 18 deletions(-)
```

### Diff
```diff
diff --git a/src/khora/memory_lake.py b/src/khora/memory_lake.py
index 5514043..54c502d 100644
--- a/src/khora/memory_lake.py
+++ b/src/khora/memory_lake.py
@@ -294,6 +294,7 @@ class MemoryLake:
         self._run_migrations = run_migrations
         self._engine: MemoryEngineProtocol | None = None
         self._connected = False
+        self._bg_tasks: set[asyncio.Task] = set()
 
     async def connect(self) -> None:
         """Connect to all storage backends."""
@@ -678,6 +679,8 @@ class MemoryLake:
         # This satisfies the durability contract — if the process crashes after
         # submit_batch() returns, the PENDING records survive for recovery.
         pending_docs: list[Document] = []
+        pending_doc_data: list[dict[str, Any]] = []
+        pre_failed_docs: list[tuple[Document, str]] = []
         for doc_data in documents:
             content = doc_data.get("content", "")
             checksum = hashlib.sha256(content.encode("utf-8")).hexdigest()
@@ -695,15 +698,24 @@ class MemoryLake:
                 extraction_config_hash=extraction_config_hash,
                 external_id=doc_data.get("external_id"),
             )
-            doc = await storage.create_document(doc)
-            pending_docs.append(doc)
+            try:
+                doc = await storage.create_document(doc)
+                pending_docs.append(doc)
+                pending_doc_data.append(doc_data)
+            except Exception as exc:
+                logger.warning(
+                    f"submit_batch: could not create document record "
+                    f"(external_id={doc_data.get('external_id')!r}): {exc}"
+                )
+                pre_failed_docs.append((doc, str(exc)))
 
-        handle = BatchHandle(batch_id=uuid4(), total=len(documents))
-        asyncio.create_task(
+        handle = BatchHandle(batch_id=uuid4(), total=len(pending_docs) + len(pre_failed_docs))
+        task = asyncio.create_task(
             self._submit_batch_worker(
                 handle,
                 pending_docs,
-                documents,
+                pending_doc_data,
+                pre_failed_docs,
                 on_result,
                 namespace_id=namespace_id,
                 skill_name=skill_name,
@@ -716,6 +728,8 @@ class MemoryLake:
                 max_concurrent=max_concurrent,
             )
         )
+        self._bg_tasks.add(task)
+        task.add_done_callback(self._bg_tasks.discard)
         return handle
 
     async def _submit_batch_worker(
@@ -723,6 +737,7 @@ class MemoryLake:
         handle: BatchHandle,
         pending_docs: list[Document],
         doc_data_list: list[dict[str, Any]],
+        pre_failed_docs: list[tuple[Document, str]],
         on_result: Callable[[int, int, DocumentResult], None],
         *,
         namespace_id: UUID,
@@ -736,20 +751,45 @@ class MemoryLake:
         max_concurrent: int,
     ) -> None:
         """Background worker that processes pre-staged PENDING documents."""
+        storage = self.storage
+
+        def _fire_result(result: DocumentResult) -> None:
+            handle._record_result(result)
+            try:
+                on_result(handle.completed, handle.total, result)
+            except Exception as cb_exc:
+                logger.warning(f"submit_batch: on_result callback raised: {cb_exc}")
+
+        # Fire error results for documents that failed to be created (e.g. external_id collision).
+        for doc, err in pre_failed_docs:
+            _fire_result(
+                DocumentResult(
+                    document_id=doc.id,
+                    namespace_id=namespace_id,
+                    success=False,
+                    error=err,
+                )
+            )
+
         engine = self._get_engine()
         process_fn = getattr(engine, "process_staged_document", None)
         if process_fn is None:
-            # Fire error results for all documents and mark handle done
+            # Engine doesn't support staged processing — mark all PENDING docs as FAILED.
+            err_msg = f"Engine {type(engine).__name__!r} does not support submit_batch (requires process_staged_document method)"
             for doc in pending_docs:
-                result = DocumentResult(
-                    document_id=doc.id,
-                    namespace_id=namespace_id,
-                    success=False,
-                    error=f"Engine {type(engine).__name__!r} does not support submit_batch "
-                    "(requires process_staged_document method)",
+                doc.mark_failed(err_msg)
+                try:
+                    await storage.update_document(doc)
+                except Exception as upd_exc:
+                    logger.warning(f"submit_batch: could not update document status: {upd_exc}")
+                _fire_result(
+                    DocumentResult(
+                        document_id=doc.id,
+                        namespace_id=namespace_id,
+                        success=False,
+                        error=err_msg,
+                    )
                 )
-                handle._record_result(result)
-                on_result(handle.completed, handle.total, result)
             handle._mark_done()
             return
 
@@ -787,17 +827,23 @@ class MemoryLake:
                     )
                 except Exception as exc:
                     logger.error(f"submit_batch: failed to process document {doc.id}: {exc}")
+                    doc.mark_failed(str(exc))
+                    try:
+                        await storage.update_document(doc)
+                    except Exception as upd_exc:
+                        logger.warning(f"submit_batch: could not update document status: {upd_exc}")
                     result = DocumentResult(
                         document_id=doc.id,
                         namespace_id=namespace_id,
                         success=False,
                         error=str(exc),
                     )
-                handle._record_result(result)
-                on_result(handle.completed, handle.total, result)
+                _fire_result(result)
 
-        await asyncio.gather(*[_process_one(doc, data) for doc, data in zip(pending_docs, doc_data_list)])
-        handle._mark_done()
+        try:
+            await asyncio.gather(*[_process_one(doc, data) for doc, data in zip(pending_docs, doc_data_list)])
+        finally:
+            handle._mark_done()
 
     async def recall(
         self,
diff --git a/tests/unit/test_memory_lake.py b/tests/unit/test_memory_lake.py
index ca312b3..8885089 100644
--- a/tests/unit/test_memory_lake.py
+++ b/tests/unit/test_memory_lake.py
@@ -1661,6 +1661,7 @@ def _make_lake_with_staged_support(ns_id):
         return doc
 
     lake._engine._storage.create_document = AsyncMock(side_effect=_fake_create_document)
+    lake._engine._storage.update_document = AsyncMock(side_effect=lambda doc: doc)
 
     async def _fake_process_staged(doc, **kwargs):
         return (2, 1, 0)  # chunks, entities, rels
@@ -1853,6 +1854,7 @@ class TestSubmitBatch:
             return doc
 
         lake._engine._storage.create_document = AsyncMock(side_effect=_fake_create)
+        lake._engine._storage.update_document = AsyncMock(side_effect=lambda doc: doc)
 
         async def _failing_process(doc, **kwargs):
             raise RuntimeError("embedding service unavailable")
@@ -1887,6 +1889,7 @@ class TestSubmitBatch:
             return doc
 
         lake._engine._storage.create_document = AsyncMock(side_effect=_fake_create)
+        lake._engine._storage.update_document = AsyncMock(side_effect=lambda doc: doc)
         # Engine has no process_staged_document attribute
         if hasattr(lake._engine, "process_staged_document"):
             del lake._engine.process_staged_document
@@ -1972,3 +1975,98 @@ class TestSubmitBatch:
         assert results[0].chunks_created == 5
         assert results[0].entities_extracted == 3
         assert results[0].relationships_created == 2
+
+    @pytest.mark.asyncio
+    async def test_failed_document_updates_storage_status(self) -> None:
+        """When process_staged_document raises, the document is marked FAILED in storage."""
+        from khora.core.models.document import DocumentStatus
+
+        ns_id = uuid4()
+        lake = _make_lake(connected=True)
+        lake._engine._storage.resolve_namespace = AsyncMock(return_value=ns_id)
+
+        async def _fake_create(doc):
+            return doc
+
+        updated_docs = []
+
+        async def _fake_update(doc):
+            updated_docs.append(doc)

... (78 lines truncated)
```

### Validation

| Metric | Count |
|--------|-------|
| Fixed (verified) | 0 |
| Fixed (unverified) | 0 |
| Accepted | 0 |
| Skipped | 0 |
| Missing evidence | 0 |
| **Total expected** | **0** |


## Audit Outcome: VERIFIED FIXED

All HIGH/MEDIUM findings from earlier cycles were verified fixed by a clean audit in cycle 2.

# Audit Findings — DYT-3054 (Cycle 2 Verification)

**Branch**: `ms/DYT-3054-khora-add-submit-batch-api`
**Commit**: `de9ac6c`
**Date**: 2026-04-24
**Type**: Verification audit — checking prior-cycle findings only

---

## Resolved Since Cycle 1

All HIGH and MEDIUM findings from cycle 1 have been fixed:

- **HIGH-1** (RESOLVED): `_process_one` now calls `doc.mark_failed(str(exc))` + `await storage.update_document(doc)` in the except block (lines 829–834). Unsupported-engine path also fixed (lines 779–784).
- **MEDIUM-1** (RESOLVED): `gather(...)` is now wrapped in `try/finally` (lines 843–846); `on_result` is wrapped in try/except inside `_fire_result` (lines 756–761).
- **MEDIUM-2** (RESOLVED): `create_document` is now wrapped in try/except (lines 701–710); failures route to `pre_failed_docs` which fires error results.
- **MEDIUM-3** (RESOLVED): `asyncio.create_task` result is now retained in `self._bg_tasks` (line 731–732; instance var at line 297).
- **TEST-MEDIUM-1** (RESOLVED): `test_failed_document_updates_storage_status` added (line 1980).
- **TEST-MEDIUM-2** (RESOLVED): `test_on_result_exception_does_not_hang` added (line 2018).
- **TEST-MEDIUM-3** (RESOLVED): `test_external_id_collision_produces_error_result` added (line 2041).

---

## Remaining Findings (LOW only)

### LOW-1: `occurred_at` key in per-document metadata undocumented

**File**: `src/khora/memory_lake.py:645–660` (docstring), `memory_lake.py:801–802`

The `submit_batch` docstring documents `metadata` as a top-level key in each document dict but does not mention that `metadata["occurred_at"]` has special temporal semantics (used as the ingestion timestamp). Callers have no way to know this without reading source.

**Still present** — docstring at line 646 reads:
```
documents: List of document dicts with 'content', 'title', 'source',
    'metadata', 'external_id' keys.
```
No mention of `occurred_at` within `metadata`.

---

### LOW-2: `content.encode("utf-8")` called twice per document

**File**: `src/khora/memory_lake.py:686, 695`

```python
checksum = hashlib.sha256(content.encode("utf-8")).hexdigest()
...
size_bytes=len(content.encode("utf-8")),
```

Encodes the same string twice. Minor inefficiency — encode once and reuse the bytes object.

**Still present**.

---

### LOW-3: `_parse_datetime` private method accessed via `getattr` duck-typing

**File**: `src/khora/memory_lake.py:803`

```python
parse_dt = getattr(engine, "_parse_datetime", None)
```

Accessing a private method via `getattr` is fragile — it can be renamed or removed without any static-analysis warning. If `_parse_datetime` is part of the extension contract for `submit_batch`, it should be documented or added to the protocol.

**Still present**.

---

### LOW-4: `process_staged_document` not declared in `MemoryEngineProtocol`

**File**: `src/khora/engines/protocol.py` (absent), `src/khora/engines/vectorcypher/engine.py:849`

`process_staged_document` is only present on `VectorCypherEngine` and is accessed via `getattr` duck-typing in the worker. It is not declared in `MemoryEngineProtocol`, so type checkers cannot verify the contract and other engine authors have no guidance on what to implement to support `submit_batch`.

**Still present** — `grep process_staged_document src/khora/engines/protocol.py` returns no matches.

---

### TEST-LOW-1: `_make_staged_doc` helper defined but never used

**File**: `tests/unit/test_memory_lake.py:1647–1652`

```python
def _make_staged_doc(ns_id):
    """Build a minimal mock Document as returned by storage.create_document."""
```

This function is defined but never called. Should be removed.

**Still present**.

---

### TEST-LOW-2: No tests for kwargs pass-through to `process_staged_document`

**File**: `tests/unit/test_memory_lake.py` — `TestSubmitBatch`

Parameters `occurred_at`, `skill_name`, `max_concurrent`, `extraction_config_hash`, and `chunk_strategy` have no assertion tests verifying they are forwarded correctly to `process_staged_document`. All `_fake_process_staged` stubs use `**kwargs` and ignore their arguments. A test using `assert_awaited_once_with` or `call_args` would prevent silent regressions.

**Still present**.

---

## Summary

| Severity | Count |
|----------|-------|
| HIGH | 0 |
| MEDIUM | 0 |
| LOW | 4 |
| Test/LOW | 2 |

**Overall**: All HIGH and MEDIUM findings from cycle 1 are resolved. Only LOW findings remain. No blocking issues.

**Comment by msrdic** [comment]
**Audits** — Fixed 5 findings, 6 LOWs noted (2 cycles)

Fixed:
- 🔴 H1: `_process_one` did not mark documents as failed on exception — fixed
- 🟡 M1: `gather(...)` not wrapped in `try/finally`; `on_result` exceptions could hang — fixed
- 🟡 M2: `create_document` failures not handled; no error result fired — fixed
- 🟡 M3: `asyncio.create_task` result not retained (fire-and-forget) — fixed
- 🟡 TM1: Missing test for failed-document storage status update — fixed
- 🟡 TM2: Missing test for `on_result` exception resilience — fixed
- 🟡 TM3: Missing test for `external_id` collision error result — fixed

Noted (advisory):
- 🟢 L1: `occurred_at` key in per-document metadata undocumented — noted
- 🟢 L2: `content.encode("utf-8")` called twice per document — noted
- 🟢 L3: `_parse_datetime` accessed via `getattr` duck-typing — noted
- 🟢 L4: `process_staged_document` not declared in `MemoryEngineProtocol` — noted
- 🟢 TL1: `_make_staged_doc` helper defined but never used — noted
- 🟢 TL2: No tests for kwargs pass-through to `process_staged_document` — noted
